File: ruby/gameplay_policy/lib/gameplay/setup_board.rb
require 'securerandom'
require_relative 'board_id'
require_relative 'board'
require_relative 'square'
require_relative 'piece'
require_relative 'board_repository'

module Gameplay
  class SetupBoard
    BACK_RANK = [Piece::Rook, Piece::Knight, Piece::Bishop, Piece::Queen,
                 Piece::King, Piece::Bishop, Piece::Knight, Piece::Rook].freeze

    def initialize(board_repository)
      @board_repository = board_repository
    end

    def call(white_player_name, black_player_name)
      board_id = BoardId.new(value: SecureRandom.uuid)
      board = Board.new(
        id: board_id,
        white_player_name: white_player_name,
        black_player_name: black_player_name,
        active_color: :white,
        pieces: initial_pieces
      )
      @board_repository.save(board)
      board_id
    end

    private

    def initial_pieces
      pieces = {}
      BACK_RANK.each_with_index do |piece_class, i|
        file = i + 1
        pieces[Square.new(file: file, rank: 1)] = piece_class.new(color: :white)
        pieces[Square.new(file: file, rank: 8)] = piece_class.new(color: :black)
        pieces[Square.new(file: file, rank: 2)] = Piece::Pawn.new(color: :white)
        pieces[Square.new(file: file, rank: 7)] = Piece::Pawn.new(color: :black)
      end
      pieces
    end
  end
end

File: python/gameplay_policy/setup_board.py
import uuid
from gameplay_policy.board_id import BoardId
from gameplay_policy.board import Board
from gameplay_policy.square import Square
from gameplay_policy.piece import King, Queen, Rook, Bishop, Knight, Pawn
from gameplay_policy.board_repository import BoardRepository

_BACK_RANK = [Rook, Knight, Bishop, Queen, King, Bishop, Knight, Rook]


def make_setup_board(board_repository: BoardRepository):
    def setup_board(white_player_name: str, black_player_name: str) -> BoardId:
        board_id = BoardId(value=str(uuid.uuid4()))
        board = Board(
            id=board_id,
            white_player_name=white_player_name,
            black_player_name=black_player_name,
            active_color="white",
            pieces=_initial_pieces(),
        )
        board_repository.save(board)
        return board_id
    return setup_board


def _initial_pieces() -> dict:
    pieces = {}
    for i, piece_class in enumerate(_BACK_RANK):
        file = i + 1
        pieces[Square(file=file, rank=1)] = piece_class(color="white")
        pieces[Square(file=file, rank=8)] = piece_class(color="black")
        pieces[Square(file=file, rank=2)] = Pawn(color="white")
        pieces[Square(file=file, rank=7)] = Pawn(color="black")
    return pieces
